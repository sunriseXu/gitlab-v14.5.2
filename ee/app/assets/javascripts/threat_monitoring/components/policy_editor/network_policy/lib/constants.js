export const RuleTypeNetwork = 'network';

export const RuleActionTypeAllow = 'allow';

export const RuleDirectionInbound = 'ingress';
export const RuleDirectionOutbound = 'egress';

export const EndpointMatchModeAny = 'any';
export const EndpointMatchModeLabel = 'label';

export const RuleTypeEndpoint = 'NetworkPolicyRuleEndpoint';
export const RuleTypeEntity = 'NetworkPolicyRuleEntity';
export const RuleTypeCIDR = 'NetworkPolicyRuleCIDR';
export const RuleTypeFQDN = 'NetworkPolicyRuleFQDN';

export const EntityTypes = {
  ALL: 'all',
  HOST: 'host',
  REMOTE_NODE: 'remote-node',
  CLUSTER: 'cluster',
  INIT: 'init',
  HEALTH: 'health',
  UNMANAGED: 'unmanaged',
  WORLD: 'world',
};

export const PortMatchModeAny = 'any';
export const PortMatchModePortProtocol = 'port/protocol';

export const DisabledByLabel = 'network-policy.gitlab.com/disabled_by';

export const ProjectIdLabel = 'app.gitlab.com/proj';
