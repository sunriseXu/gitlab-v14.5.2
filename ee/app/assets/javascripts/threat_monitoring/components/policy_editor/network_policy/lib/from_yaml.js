import { safeLoad } from 'js-yaml';
import {
  DisabledByLabel,
  EndpointMatchModeAny,
  EndpointMatchModeLabel,
  RuleDirectionInbound,
  RuleDirectionOutbound,
  PortMatchModeAny,
  PortMatchModePortProtocol,
  RuleTypeEndpoint,
  RuleTypeEntity,
  RuleTypeCIDR,
  RuleTypeFQDN,
} from './constants';
import { buildRule } from './rules';

/*
  Convert list of matchLabel selectors used by the endpoint rule to an
  entity rule object expected by the rule builder.

  We expect list of object in format:
  [{ matchLabels: { foo: 'bar' } }, { matchLabels: { bar: 'baz' } }]
  And will return a single rule object:
  { matchLabels: 'foo:bar baz:bar' }
*/
function ruleTypeEndpointFunc(items) {
  const labels = items
    .reduce(
      (acc, { matchLabels }) =>
        acc.concat(Object.keys(matchLabels).map((key) => `${key}:${matchLabels[key]}`)),
      [],
    )
    .join(' ');
  return { matchLabels: labels };
}

function ruleTypeEntityFunc(entities) {
  return { entities };
}

function ruleTypeCIDRFunc(items) {
  const cidr = items.join(' ');
  return { cidr };
}

/*
  Convert list of matchName selectors used by the fqdn rule to a
  fqdn rule object expected by the rule builder.

  We expect list of object in format:
  [{ matchName: 'remote-service.com' }, { matchName: 'another-service.com' }]
  And will return a single rule object:
  { fqdn: 'remote-service.com another-service.com' }
*/
function ruleTypeFQDNFunc(items) {
  const fqdn = items.map(({ matchName }) => matchName).join(' ');
  return { fqdn };
}

const rulesFunc = {
  [RuleTypeEndpoint]: ruleTypeEndpointFunc,
  [RuleTypeEntity]: ruleTypeEntityFunc,
  [RuleTypeCIDR]: ruleTypeCIDRFunc,
  [RuleTypeFQDN]: ruleTypeFQDNFunc,
};

/*
  Parse yaml rule into an object expected by the policy editor.
*/
function parseRule(item, direction) {
  let ruleItem;
  let ruleType;

  if (item.fromEntities || item.toEntities) {
    ruleType = RuleTypeEntity;
    ruleItem = item.fromEntities || item.toEntities;
  } else if (item.fromCIDR || item.toCIDR) {
    ruleType = RuleTypeCIDR;
    ruleItem = item.fromCIDR || item.toCIDR;
  } else if (item.toFQDNs) {
    ruleType = RuleTypeFQDN;
    ruleItem = item.toFQDNs;
  } else {
    ruleItem = item.fromEndpoints || item.toEndpoints || [];
    ruleType = RuleTypeEndpoint;
  }

  let portMatchMode = PortMatchModeAny;
  let portList = [];
  if (item.toPorts?.length > 0) {
    portMatchMode = PortMatchModePortProtocol;
    portList = item.toPorts.reduce(
      (acc, { ports }) =>
        acc.concat(ports.map(({ port, protocol = 'TCP' }) => `${port}/${protocol.toLowerCase()}`)),
      [],
    );
  }

  return {
    ...buildRule(ruleType, {
      direction,
      portMatchMode,
      ports: portList.join(' '),
    }),
    ...rulesFunc[ruleType](ruleItem),
  };
}

/**
 * Checks for parameters unsupported by the network policy "Rule Mode"
 * @param {String} manifest YAML of network policy
 * @returns {Boolean} whether the YAML is valid to be parsed into "Rule Mode"
 */
const hasUnsupportedAttribute = (manifest) => {
  const primaryKeys = ['apiVersion', 'description', 'kind', 'metadata', 'spec'];
  const metadataKeys = ['annotations', 'labels', 'name', 'namespace', 'resourceVersion'];
  const specKeys = ['egress', 'endpointSelector', 'ingress'];
  const ruleKeys = [
    'fromEntities',
    'toEntities',
    'fromCIDR',
    'toCIDR',
    'toFQDNs',
    'fromEndpoints',
    'toEndpoints',
    'toPorts',
  ];
  const toPortKeys = ['ports'];
  const portKeys = ['port', 'protocol'];
  let isUnsupported = false;
  const hasInvalidKey = (object, allowedValues) => {
    return !Object.keys(object).every((item) => allowedValues.includes(item));
  };

  const hasInvalidPolicy = (ingress = [], egress = []) => {
    let isInvalidPolicy = false;
    [...ingress, ...egress].forEach((item) => {
      isInvalidPolicy = hasInvalidKey(item, ruleKeys);
      if (item.toPorts?.length && !isInvalidPolicy) {
        item.toPorts.forEach((entry) => {
          isInvalidPolicy = hasInvalidKey(entry, toPortKeys);
          if (entry.ports?.length && !isInvalidPolicy) {
            entry.ports.forEach((portEntry) => {
              isInvalidPolicy = hasInvalidKey(portEntry, portKeys);
            });
          }
        });
      }
    });
    return isInvalidPolicy;
  };

  isUnsupported = hasInvalidKey(manifest, primaryKeys);

  if (manifest?.metadata && !isUnsupported) {
    isUnsupported = hasInvalidKey(manifest.metadata, metadataKeys);
  }
  if (manifest?.spec && !isUnsupported) {
    isUnsupported = hasInvalidKey(manifest.spec, specKeys);
  }
  if (!isUnsupported && (manifest?.spec?.ingress?.length || manifest?.spec?.egress?.length)) {
    isUnsupported = hasInvalidPolicy(manifest.spec.ingress, manifest.spec.egress);
  }

  return isUnsupported;
};

/*
  Construct a policy object expected by the policy editor from a yaml manifest.
  Expected yaml structure is defined in the official documentation:
    https://docs.cilium.io/en/v1.8/policy/language
*/
export default function fromYaml(manifest) {
  const policy = safeLoad(manifest, { json: true });

  const unsupportedAttribute = hasUnsupportedAttribute(policy);
  if (unsupportedAttribute) return { error: unsupportedAttribute };

  const { description, metadata, spec } = policy;
  const { name, resourceVersion, annotations, labels } = metadata;
  const { endpointSelector = {}, ingress = [], egress = [] } = spec;
  const matchLabels = endpointSelector.matchLabels || {};

  const endpointLabels = Object.keys(matchLabels).reduce((acc, key) => {
    if (key === DisabledByLabel) return acc;
    acc.push(`${key}:${matchLabels[key]}`);
    return acc;
  }, []);

  const rules = []
    .concat(
      ingress.map((item) => parseRule(item, RuleDirectionInbound)),
      egress.map((item) => parseRule(item, RuleDirectionOutbound)),
    )
    .filter((rule) => Boolean(rule));

  return {
    name,
    resourceVersion,
    description,
    annotations,
    labels,
    isEnabled: !Object.keys(matchLabels).includes(DisabledByLabel),
    endpointMatchMode: endpointLabels.length > 0 ? EndpointMatchModeLabel : EndpointMatchModeAny,
    endpointLabels: endpointLabels.join(' '),
    rules,
  };
}
