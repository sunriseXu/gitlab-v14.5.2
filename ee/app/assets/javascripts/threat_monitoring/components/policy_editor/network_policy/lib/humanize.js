import { sprintf, __, s__ } from '~/locale';
import {
  EndpointMatchModeAny,
  RuleDirectionInbound,
  PortMatchModeAny,
  RuleTypeEntity,
  RuleTypeCIDR,
  RuleTypeFQDN,
} from './constants';
import { portSelectors, labelSelector, splitItems } from './utils';

const strongArgs = { strongOpen: '<strong>', strongClose: '</strong>' };

/*
 Return humanizied description for a port matcher of a rule.
*/
function humanizeNetworkPolicyRulePorts(rule) {
  const { portMatchMode } = rule;

  if (portMatchMode === PortMatchModeAny)
    return sprintf(s__('NetworkPolicies|%{strongOpen}any%{strongClose} port'), strongArgs, false);

  const portList = portSelectors(rule);
  const ports = portList.map(({ port, protocol }) => `${port}/${protocol}`).join(', ');
  return sprintf(
    s__('NetworkPolicies|ports %{ports}'),
    {
      ports: `<strong>${ports}</strong>`,
    },
    false,
  );
}

/*
 Return humanizied description of an endpoint rule.
*/
function humanizeNetworkPolicyRuleEndpoint({ matchLabels }) {
  const matchSelector = labelSelector(matchLabels);
  const labels = Object.keys(matchSelector)
    .map((key) => `${key}: ${matchSelector[key]}`)
    .join(', ');
  return labels.length === 0
    ? sprintf(s__('NetworkPolicies|%{strongOpen}all%{strongClose} pods'), strongArgs, false)
    : sprintf(
        s__('NetworkPolicies|pods %{pods}'),
        {
          pods: `<strong>[${labels}]</strong>`,
        },
        false,
      );
}

/*
 Return humanizied description of an entity rule.
*/
function humanizeNetworkPolicyRuleEntity({ entities }) {
  const entitiesList = entities.length === 0 ? s__('NetworkPolicies|nowhere') : entities.join(', ');
  return `<strong>${entitiesList}</strong>`;
}

/*
 Return humanizied description of a cidr rule.
*/
function humanizeNetworkPolicyRuleCIDR({ cidr }) {
  const cidrList = splitItems(cidr);
  const cidrs =
    cidrList.length === 0 ? s__('NetworkPolicies|all IP addresses') : cidrList.join(', ');
  return `<strong>${cidrs}</strong>`;
}

/*
 Return humanizied description of a fqdn rule.
*/
function humanizeNetworkPolicyRuleFQDN({ fqdn }) {
  const fqdnList = splitItems(fqdn);
  const fqdns = fqdnList.length === 0 ? s__('NetworkPolicies|all DNS names') : fqdnList.join(', ');
  return `<strong>${fqdns}</strong>`;
}

/*
 Return humanizied description of a rule.
*/
function humanizeNetworkPolicyRule(rule) {
  const { ruleType } = rule;

  switch (ruleType) {
    case RuleTypeEntity:
      return humanizeNetworkPolicyRuleEntity(rule);
    case RuleTypeCIDR:
      return humanizeNetworkPolicyRuleCIDR(rule);
    case RuleTypeFQDN:
      return humanizeNetworkPolicyRuleFQDN(rule);
    default:
      return humanizeNetworkPolicyRuleEndpoint(rule);
  }
}

/*
 Return humanizied description of an endpoint matcher of a policy.
*/
function humanizeEndpointSelector({ endpointMatchMode, endpointLabels }) {
  if (endpointMatchMode === EndpointMatchModeAny)
    return sprintf(s__('NetworkPolicies|%{strongOpen}all%{strongClose} pods'), strongArgs, false);

  const selector = labelSelector(endpointLabels);
  const pods = Object.keys(selector)
    .map((key) => `${key}: ${selector[key]}`)
    .join(', ');
  return sprintf(
    s__('NetworkPolicies|pods %{pods}'),
    {
      pods: `<strong>[${pods}]</strong>`,
    },
    false,
  );
}

/*
 Return humanizied description of a provided network policy.
*/
export default function humanizeNetworkPolicy(policy) {
  const { rules } = policy;
  if (rules.length === 0) return s__('NetworkPolicies|Deny all traffic');

  const selector = humanizeEndpointSelector(policy);

  const humanizedRules = rules.map((rule) => {
    const { direction } = rule;
    const template =
      direction === RuleDirectionInbound
        ? s__(
            'NetworkPolicies|Allow all inbound traffic to %{selector} from %{ruleSelector} on %{ports}',
          )
        : s__(
            'NetworkPolicies|Allow all outbound traffic from %{selector} to %{ruleSelector} on %{ports}',
          );
    const ruleSelector = humanizeNetworkPolicyRule(rule);
    const ports = humanizeNetworkPolicyRulePorts(rule);
    return sprintf(template, { selector, ruleSelector, ports }, false);
  });

  return humanizedRules.join(`<br><br>${__('and').toLocaleUpperCase()}<br><br>`);
}
