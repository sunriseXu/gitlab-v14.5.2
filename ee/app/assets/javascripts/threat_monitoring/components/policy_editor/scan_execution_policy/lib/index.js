export { fromYaml } from './from_yaml';
export { toYaml } from './to_yaml';

export * from './constants';
export * from './humanize';
export * from './utils';

export const DEFAULT_SCAN_EXECUTION_POLICY = `type: scan_execution_policy
name: ''
description: ''
enabled: false
rules:
  - type: pipeline
    branches:
      - main
actions:
  - scan: dast
    site_profile: ''
    scanner_profile: ''
`;
