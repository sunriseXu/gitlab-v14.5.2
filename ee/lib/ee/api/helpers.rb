# frozen_string_literal: true

module EE
  module API
    module Helpers
      extend ::Gitlab::Utils::Override
      include ::Gitlab::Utils::StrongMemoize

      def require_node_to_be_enabled!
        forbidden! 'Geo node is disabled.' unless ::Gitlab::Geo.current_node&.enabled?
      end

      def gitlab_geo_node_token?
        headers['Authorization']&.start_with?(::Gitlab::Geo::BaseRequest::GITLAB_GEO_AUTH_TOKEN_TYPE)
      end

      def authenticate_by_gitlab_geo_node_token!
        unauthorized! unless authorization_header_valid?
      rescue ::Gitlab::Geo::InvalidDecryptionKeyError, ::Gitlab::Geo::InvalidSignatureTimeError => e
        render_api_error!(e.to_s, 401)
      end

      def geo_jwt_decoder
        return unless gitlab_geo_node_token?

        strong_memoize(:geo_jwt_decoder) do
          ::Gitlab::Geo::JwtRequestDecoder.new(headers['Authorization'])
        end
      end

      # Update the jwt_decoder to allow authorization of disabled (paused) nodes
      def allow_paused_nodes!
        geo_jwt_decoder.include_disabled!
      end

      def check_gitlab_geo_request_ip!
        unauthorized! unless ::Gitlab::Geo.allowed_ip?(request.ip)
      end

      def authorization_header_valid?
        return unless gitlab_geo_node_token?

        scope = geo_jwt_decoder.decode.try { |x| x[:scope] }
        scope == ::Gitlab::Geo::API_SCOPE
      end

      def check_project_feature_available!(feature)
        not_found! unless user_project.feature_available?(feature)
      end

      def authorize_change_param(subject, *keys)
        keys.each do |key|
          authorize!("change_#{key}".to_sym, subject) if params.has_key?(key)
        end
      end

      # Normally, only admin users should have access to see LDAP
      # groups. However, due to the "Allow group owners to manage LDAP-related
      # group settings" setting, any group owner can sync LDAP groups with
      # their project.
      #
      # In the future, we should also check that the user has access to manage
      # a specific group so that we can use the Ability class.
      def authenticated_with_ldap_admin_access!
        authenticate!

        forbidden! unless current_user.admin? ||
            ::Gitlab::CurrentSettings.current_application_settings
              .allow_group_owners_to_manage_ldap
      end

      override :find_project!
      def find_project!(id)
        project = find_project(id)

        return forbidden! unless authorized_project_scope?(project)

        # CI job token authentication:
        # this method grants limited privileged for admin users
        # admin users can only access project if they are direct member
        ability = job_token_authentication? ? :build_read_project : :read_project

        return project if can?(current_user, ability, project)
        return unauthorized! if authenticate_non_public?

        not_found!('Project')
      end

      override :find_group!
      def find_group!(id)
        # CI job token authentication:
        # currently we do not allow any group access for CI job token
        if job_token_authentication?
          not_found!('Group')
        else
          super
        end
      end

      # rubocop: disable CodeReuse/ActiveRecord
      def find_group_epic(iid)
        EpicsFinder.new(current_user, group_id: user_group.id).find_by!(iid: iid)
      end
      # rubocop: enable CodeReuse/ActiveRecord

      private

      override :project_finder_params_ee
      def project_finder_params_ee
        if params[:with_security_reports].present?
          { with_security_reports: true }
        else
          {}
        end
      end

      override :send_git_archive
      def send_git_archive(repository, **kwargs)
        AuditEvents::RepositoryDownloadStartedAuditEventService.new(
          current_user,
          repository.project,
          ip_address
        ).for_project.security_event

        super
      end

      def private_token
        params[::APIGuard::PRIVATE_TOKEN_PARAM] || env[::APIGuard::PRIVATE_TOKEN_HEADER]
      end

      def warden
        env['warden']
      end

      # Check if the request is GET/HEAD, or if CSRF token is valid.
      def verified_request?
        ::Gitlab::RequestForgeryProtection.verified?(env)
      end

      # Check the Rails session for valid authentication details
      def find_user_from_warden
        warden.try(:authenticate) if verified_request?
      end

      def geo_token
        ::Gitlab::Geo.current_node.system_hook.token
      end

      def authorize_manage_saml!(group)
        unauthorized! unless can?(current_user, :admin_group_saml, group)
      end

      def check_group_saml_configured
        forbidden!('Group SAML not enabled.') unless ::Gitlab::Auth::GroupSaml::Config.enabled?
      end
    end
  end
end
