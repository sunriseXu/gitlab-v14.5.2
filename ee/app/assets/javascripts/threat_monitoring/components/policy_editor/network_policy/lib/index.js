import { EndpointMatchModeAny } from './constants';

export * from './constants';
export { default as fromYaml } from './from_yaml';
export { default as humanizeNetworkPolicy } from './humanize';
export { buildRule } from './rules';
export { default as toYaml } from './to_yaml';

export const DEFAULT_NETWORK_POLICY = {
  name: '',
  description: '',
  isEnabled: false,
  endpointMatchMode: EndpointMatchModeAny,
  endpointLabels: '',
  annotations: '',
  labels: '',
};
