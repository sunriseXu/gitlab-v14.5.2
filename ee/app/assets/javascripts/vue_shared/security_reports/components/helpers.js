import { EMPTY_BODY_MESSAGE } from './constants';

/**
 * A helper function which validates the passed
 * in body string.
 *
 * It returns an empty string if the body has explicitly
 * been passed in as an empty string, a fallback
 * message if the body is null / undefined, else
 * it will return the original body string.
 *
 * @param {String} body the body message
 *
 * @return {String} the validated body message
 */
export const bodyWithFallBack = (body) => (body === '' ? '' : body || EMPTY_BODY_MESSAGE);
