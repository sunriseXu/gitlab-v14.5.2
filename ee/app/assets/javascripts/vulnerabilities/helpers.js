import { getIdFromGraphQLId } from '~/graphql_shared/utils';
import { isAbsolute, isSafeURL } from '~/lib/utils/url_utility';
import { REGEXES } from './constants';

// Get the issue in the format expected by the descendant components of related_issues_block.vue.
export const getFormattedIssue = (issue) => ({
  ...issue,
  reference: `#${issue.iid}`,
  path: issue.web_url,
});

export const getAddRelatedIssueRequestParams = (reference, defaultProjectId) => {
  let issueId = reference;
  let projectId = defaultProjectId;

  // If the reference is an issue number, parse out just the issue number.
  if (REGEXES.ISSUE_FORMAT.test(reference)) {
    [, issueId] = REGEXES.ISSUE_FORMAT.exec(reference);
  }
  // If the reference is an absolute URL and matches the issues URL format, parse out the project and issue.
  else if (isSafeURL(reference) && isAbsolute(reference)) {
    const { pathname } = new URL(reference);

    if (REGEXES.LINK_FORMAT.test(pathname)) {
      [, projectId, issueId] = REGEXES.LINK_FORMAT.exec(pathname);
    }
  }

  return { target_issue_iid: issueId, target_project_id: projectId };
};

export const normalizeGraphQLNote = (note) => {
  if (!note) {
    return null;
  }

  return {
    ...note,
    id: getIdFromGraphQLId(note.id),
    author: {
      ...note.author,
      id: getIdFromGraphQLId(note.author.id),
      path: note.author.webPath,
    },
  };
};

export const normalizeGraphQLVulnerability = (vulnerability) => {
  if (!vulnerability) {
    return null;
  }

  const newVulnerability = { ...vulnerability };

  if (vulnerability.id) {
    newVulnerability.id = getIdFromGraphQLId(vulnerability.id);
  }

  if (vulnerability.state) {
    newVulnerability.state = vulnerability.state.toLowerCase();
  }

  ['confirmed', 'resolved', 'dismissed'].forEach((state) => {
    if (vulnerability[`${state}By`]?.id) {
      newVulnerability[`${state}ById`] = getIdFromGraphQLId(vulnerability[`${state}By`].id);
      delete newVulnerability[`${state}By`];
    }
  });

  return newVulnerability;
};
