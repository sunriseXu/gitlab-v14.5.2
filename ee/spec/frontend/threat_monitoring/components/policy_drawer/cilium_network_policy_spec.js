import { GlIntersperse } from '@gitlab/ui';
import BasePolicy from 'ee/threat_monitoring/components/policy_drawer/base_policy.vue';
import CiliumNetworkPolicy from 'ee/threat_monitoring/components/policy_drawer/cilium_network_policy.vue';
import { toYaml } from 'ee/threat_monitoring/components/policy_editor/network_policy/lib';
import PolicyPreview from 'ee/threat_monitoring/components/policy_editor/policy_preview.vue';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';

describe('CiliumNetworkPolicy component', () => {
  let wrapper;
  const supportedYaml = toYaml({
    name: 'test-policy',
    description: 'test description',
    endpointLabels: '',
    rules: [],
  });
  const unsupportedYaml = 'unsupportedPrimaryKey: test';

  const findPolicyPreview = () => wrapper.findComponent(PolicyPreview);
  const findDescription = () => wrapper.findByTestId('description');
  const findEnvironments = () => wrapper.findByTestId('environments');

  const factory = ({ propsData } = {}) => {
    wrapper = shallowMountExtended(CiliumNetworkPolicy, {
      propsData: {
        ...propsData,
      },
      stubs: {
        BasePolicy,
        GlIntersperse,
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
  });

  describe('supported YAML', () => {
    beforeEach(() => {
      factory({ propsData: { policy: { yaml: supportedYaml } } });
    });

    it('renders policy preview tabs', () => {
      expect(wrapper.find('div').element).toMatchSnapshot();
    });

    it('does render the policy description', () => {
      expect(findDescription().exists()).toBe(true);
      expect(findDescription().text()).toContain('test description');
    });

    it('does render the policy preview', () => {
      expect(findPolicyPreview().exists()).toBe(true);
      expect(findPolicyPreview().props()).toStrictEqual({
        initialTab: 0,
        policyDescription: 'Deny all traffic',
        policyYaml: supportedYaml,
      });
    });
  });

  describe('unsupported YAML', () => {
    beforeEach(() => {
      factory({ propsData: { policy: { yaml: unsupportedYaml } } });
    });

    it('renders policy preview tabs', () => {
      expect(wrapper.find('div').element).toMatchSnapshot();
    });

    it('does not render the policy description', () => {
      expect(findDescription().exists()).toBe(false);
    });

    it('does render the policy preview', () => {
      expect(findPolicyPreview().exists()).toBe(true);
      expect(findPolicyPreview().props()).toStrictEqual({
        initialTab: 1,
        policyDescription: null,
        policyYaml: unsupportedYaml,
      });
    });
  });

  describe('environments', () => {
    it('renders environments if any', () => {
      factory({
        propsData: {
          policy: {
            environments: {
              nodes: [{ name: 'production' }, { name: 'local' }],
            },
            yaml: supportedYaml,
          },
        },
      });
      expect(findEnvironments().exists()).toBe(true);
      expect(findEnvironments().text()).toContain('production');
      expect(findEnvironments().text()).toContain('local');
    });

    it("does not render environments row if there aren't any", () => {
      factory({
        propsData: {
          policy: {
            environments: {
              nodes: [],
            },
            yaml: supportedYaml,
          },
        },
      });
      expect(findEnvironments().exists()).toBe(false);
    });
  });
});
