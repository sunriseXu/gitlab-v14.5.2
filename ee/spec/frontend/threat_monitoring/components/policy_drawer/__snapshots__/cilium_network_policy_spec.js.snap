// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CiliumNetworkPolicy component supported YAML renders policy preview tabs 1`] = `
<div>
  <h5
    class="gl-mt-3"
  >
    Type
  </h5>
   
  <p
    data-testid="policy-type"
  >
    Network
  </p>
   
  <div>
    <policy-info-row-stub
      data-testid="description"
      label="Description"
    >
      
        test description
      
    </policy-info-row-stub>
     
    <policy-info-row-stub
      label="Status"
    >
      Disabled
    </policy-info-row-stub>
     
    <!---->
  </div>
   
  <policy-preview-stub
    class="gl-mt-4"
    initialtab="0"
    policydescription="Deny all traffic"
    policyyaml="apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
description: test description
metadata:
  name: test-policy
spec:
  endpointSelector:
    matchLabels:
      network-policy.gitlab.com/disabled_by: gitlab
"
  />
</div>
`;

exports[`CiliumNetworkPolicy component unsupported YAML renders policy preview tabs 1`] = `
<div>
  <h5
    class="gl-mt-3"
  >
    Type
  </h5>
   
  <p
    data-testid="policy-type"
  >
    Network
  </p>
   
  <!---->
   
  <policy-preview-stub
    class="gl-mt-4"
    initialtab="1"
    policyyaml="unsupportedPrimaryKey: test"
  />
</div>
`;
